### PR TITLE
Add user in widget 2.0

### PR DESCRIPTION
**Pull request by chrisnojima**
- [x] Moved popup-menu to common component
- [x] Implement add user button

### CODE REVIEW

File: shared/actions/kbfs/index.desktop.js [hunk 1]
```diff
@@ -4,8 +4,19 @@ import {shell} from 'electron'
 import * as Constants from '../../constants/config'
 import {config} from '../../constants/types/keybase-v1'
 import type {AsyncAction} from '../../constants/types/flux'
+import {cleanup} from '../../util/kbfs'
 
 const open = (kbfsPath: string, path: ?string) => {
```
Comment (chrisnojima) [inline shared/actions/kbfs/index.desktop.js]: ensure we call cleanup on all paths that come in. the current cleanup assumes you don't pass the /private /public prefix so this'll strip that out then re-add it

File: shared/common-adapters/button.desktop.js [hunk 2]
```diff
@@ -117,7 +117,7 @@ class Button extends Component {
         <FlatButton
           onClick={this.props.onClick}
           style={{...backgroundStyle, ...smallStyle, ...this.props.style}}
-          labelStyle={{...stylesButtonLabel, ...labelStyle}}
```
Comment (chrisnojima) [inline shared/common-adapters/button.desktop.js]: giving me some extra styling hooks, same on input below

File: shared/common-adapters/popup-menu.desktop.js [hunk 3]
```diff
@@ -0,0 +1,62 @@
+// @flow
+import React, {Component} from 'react'
+import {Box, Text} from '../common-adapters/index'
+import type {Props} from './popup-menu'
+import {globalColors, globalStyles} from '../styles/style-guide'
+
+class Menu extends Component<void, Props, void> {
```
Comment (chrisnojima) [inline shared/common-adapters/popup-menu.desktop.js]: pulled the popup menu out. kept it mostly the same. we can expand it later

File: shared/constants/keycodes.js [hunk 4]
```diff
@@ -0,0 +1,116 @@
+// @flow
+
```
Comment (chrisnojima) [inline shared/constants/keycodes.js]: we can use this for keycodes

File: shared/folders/dumb.js [hunk 5]
```diff
@@ -129,7 +129,7 @@ export const map: DumbComponentMap<Folders> = {
   mocks: {
     'Normal': {
       private: {tlfs, ignored, isPublic: false, parentProps},
-      public: {tlfs, ignored, isPublic: true, privateBadge: 1, publicBadge: 222, parentProps}
+      public: {tlfs: [f2, f3, f4, f5], ignored, isPublic: true, privateBadge: 1, publicBadge: 222, parentProps}
```
Comment (chrisnojima) [inline shared/folders/dumb.js]: make a slightly different list so its more clear its working

File: shared/folders/list.desktop.js [hunk 6]
```diff
@@ -58,6 +58,7 @@ class Render extends Component<void, Props, State> {
       <Box style={{...stylesScrollContainer, ...this.props.style}}>
         <Box style={stylesContainer}>
           <style>{realCSS}</style>
+          {this.props.extraRows}
```
Comment (chrisnojima) [inline shared/folders/list.desktop.js]: allow us to add extra rows into the folder list

File: shared/menubar/user-add.desktop.js [hunk 7]
```diff
@@ -0,0 +1,141 @@
+import React, {Component} from 'react'
```
Comment (chrisnojima) [inline shared/menubar/user-add.desktop.js]: shows a button or an input

File: shared/reducers/favorite.js [hunk 8]
```diff
@@ -16,7 +17,7 @@ const initialState = {
 
 const folderToProps = (folders, username) => {
   const converted = folders.map(f => ({
-    users: f.name.split(',').map(u => ({
+    users: canonicalizeUsernames(username || '', parseFolderNameToUsers(f.name)).map(u => ({
```
Comment (chrisnojima) [inline shared/reducers/favorite.js]: we should use these helpers always

**Comment by chrisnojima** [comment]
screen shot 2016-05-12 at 3 11 59 pm
screen shot 2016-05-12 at 3 11 52 pm
screen shot 2016-05-12 at 3 11 42 pm
screen shot 2016-05-12 at 3 11 38 pm
screen shot 2016-05-12 at 3 11 32 pm
screen shot 2016-05-12 at 3 11 26 pm
screen shot 2016-05-12 at 3 11 15 pm

**Comment by chrisnojima** [comment]
@keybase/react-hackers this is ready to take a look at

File: shared/actions/kbfs/index.desktop.js [hunk 9]
```diff
@@ -4,15 +4,27 @@ import {shell} from 'electron'
...
+  path = path || ''
```
Comment (chromakode) [inline shared/actions/kbfs/index.desktop.js]: es6 default parameter? :grin:

**Comment by chromakode** [comment]
:+1:

File: shared/actions/kbfs/index.desktop.js (hunk 9 above)
Comment (chrisnojima) [inline shared/actions/kbfs/index.desktop.js]: annoyingly we have to put an eslint thing in there as it'll complain i think

File: shared/menubar/user-add.desktop.js (hunk 7 above)
Comment (MarcoPolo) [inline shared/menubar/user-add.desktop.js]: this isn't flow typed, but you're using flow types

File: shared/menubar/index.js [hunk 10]
```diff
@@ -18,12 +18,12 @@ import type {Props as RenderProps} from './index.render'
...
 export type Props = {
   username: ?string,
-  folders: ?RenderProps,
   favoriteList: () => void,
   openInKBFS: (target?: any) => void,
   loggedIn: ?boolean,
-  switchTab: (tab: string) => void
-} & RenderProps
+  switchTab: (tab: string) => void,
+  folderProps: ?RenderProps
```
Comment (MarcoPolo) [inline shared/menubar/index.js]: This should be FolderProps

**Comment by MarcoPolo** [comment]
👍 % C

**Pull request merged**